### ISSUE TITLE
Application hangs when trying to run it under Linux Debian Buster with xfce4 desktop

### ISSUE THREAD

**Issue by tddpirate** [ImpulseAdventure/GUIslice-Builder#57]
**Describe the bug**
After installation of GUIslice Builder from `builder-linux-0.13.b011.tar.gz` into `/opt/GUIsliceBuilder`, when trying to run it under Linux Debian Buster with xfce4 desktop (version 4.12.5) using Java from Arduino 1.8.10: the application throws all the time the exception:
`Exception: java.lang.NullPointerException thrown from the UncaughtExceptionHandler in thread "AWT-EventQueue-0"`
and the application's window does not execute when I try to click on menu items. I have to abort it using Ctrl-C.

**To Reproduce**
1. Install the application as instructed above (Linux Debian Buster (Debian 10.2)), xfce4, `/opt/GUIsliceBuilder`, Use Java from Arduino 1.8.10 (in my case, installed in: `/opt/arduino/arduino-1.8.10`).
2. Open a terminal (in my case, it was xfce4-terminal 0.8.7.4).
3. `cd /opt/GUIsliceBuilder`
4. `./GUIslice.sh` (without `&` so that you can abort it by entering Ctrl-C on the terminal window).

**Expected behavior**
A responsive application window. In particular, I can close it by clicking on File / Exit rather by aborting it from the terminal window.

**Version**
- Builder version: 0.13.b011
- OS: Linux Debian Buster (10.2)

**Comment by tddpirate** [comment]
After uninstallation and reinstallation into my home directory, the application crashed. Crash log is attached.
[20200108_005406.txt](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4032839/20200108_005406.txt)

**Comment by Pconti31** [comment]
Basically, the crash file is saying the app read your ~/GUIsliceBuilder/templates/arduinofonts.csv file and found it to be zero length.  Causing the crash.  I would understand if it couldn't find the file but zero length is Weird.

You say you ran guislice.sh in an arbitrary directory?  did you try inside your ~/GUISliceBuilder directory?

AS to the first issue, the doc clearly state to install in your home directory.  It's not designed to be shared among many users.  That said the only time I found the screens to be unresponsive is when canberra-gtk-module was not up to date.   Running this command from the the terminal
`sudo apt install libcanberra-gtk-module libcanberra-gtk3-modulelib`
fixed that for me. I use standard ubuntu 18.04 so it could be an interaction with your desktop software causing screen freeze.

Please post your guislice.sh file so I can look at it and if you haven't already try running it from GUIsliceBuilder directory.  What version of java are you running? 
Also, try and open to arduinofonts.cvs file with your text editor and see if anything looks broken.
LIke zero length file or or end of line not being respected.   
Paul--

**Comment by tddpirate** [comment]
Paul,
Thanks for your response.

After reinstallation (see below), `~/GUIsliceBuilder/templates/arduinofonts.csv` is 4974 bytes long and seems to be OK (it has CRLF line terminators).

I re-installed GUIsliceBuilder at my home directory. After `cd ~/GUIsliceBuilder`, I did `./GUIslice.sh` and the application did not crash but the screens were unresponsive.
The following error message repeated several times:
`Exception: java.lang.NullPointerException thrown from the UncaughtExceptionHandler in thread "AWT-EventQueue-0"`

My system has:
`libcanberra-gtk-module`                 0.30-7
`libcanberra-gtk3-module`               0.30-7
Debian Buster does not have `libcanberra-gtk3-modulelib`.

Attached please find `GUIslice.sh` (renamed as `GUIslice.sh.txt` because GitHub does not let `*.sh` files be uploaded).
[GUIslice.sh.txt](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4036231/GUIslice.sh.txt)

My GUIsliceBuilder was configured to run with Arduino 1.8.10 supplied Java version:
```
$ /opt/arduino/arduino-1.8.10/java/bin/java -version
java version "1.8.0_191"
Java(TM) SE Runtime Environment (build 1.8.0_191-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.191-b12, mixed mode)
```

**Comment by Pconti31** [comment]
Ok, i'm running libcanberra-gtk 0.30-5 but I don't think that's the problem.
I built a special version of builder-0.13.b011.jar for you that doesn't trap unhandled exceptions.
Hopefully we will get a crash file with more info.   simply replace the jar and run the guislice.sh from a terminal in case we get any console output.
Paul--
[builder-0.13.b011.jar.zip](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4036683/builder-0.13.b011.jar.zip)

**Comment by tddpirate** [comment]
I replaced the `builder-0.13.b011.jar` file and re-ran `./GUIslice.sh`.
The application windows continued to be unresponsive, until I aborted the application by Ctrl-C.
There was no console output at all.

I now notice that the `-classpath` argument to the java command running `builder.Builder` quotes `*.jar` files which I do not have:
`/home/omer/GUIsliceBuilder/lib/flamingo-7.3.1-SNAPSHOT.jar
/home/omer/GUIsliceBuilder/lib/laf-plugin-7.3.1-SNAPSHOT.jar
/home/omer/GUIsliceBuilder/lib/laf-widget-7.3.1-SNAPSHOT.jar
/home/omer/GUIsliceBuilder/lib/substance-7.3.1-SNAPSHOT.jar
/home/omer/GUIsliceBuilder/lib/trident-7.3.1-SNAPSHOT.jar`
Inside your `builder-0.13.b011.jar` I notice that there are classes for `flamingo`, `lafplugin`, `lafwidget`, `substance`, `trident`.

Could it be that the jar file has pre-7.3.1 versions of those classes, and they were overridden in your setup but not in my setup?

**Comment by Pconti31** [comment]
Saddy No.  While I should remove them from the classpath and you can do so if you want Those jars are included inside the builder-0.13.b011.jar file.  It's called a FAT jar and makes it easier to support macs. 
Did a crash file get created inside GUIsliceBuilder/logs? 
Paul--

**Comment by tddpirate** [comment]
When I said that the application was not responsive, it was not 100% true. So I decided to capture the screen when starting and stopping it. A video is attached. Unzip before playing it.

You'll notice that some controls are shown only when the mouse goes over them.
There is a control which responds to mouse click by closing a page.
There is an indication that pages get the focus when clicked on.
The File menu item does not open its submenu so that I could not close the application and had to abort it using Ctrl-C instead (there is no crash logfile whatsoever).

Hopefully it'll help you understand what is going on.

[Kazam_screencast_00000.zip](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4036952/Kazam_screencast_00000.zip)

**Comment by Pconti31** [comment]
Actually, this is the same behavior I saw when libcanberra-gtk-modules were incorrect.
However in that case I got a missing library console message so that can't be the problem.

I made another jar for you this one adds more exception traps.
Seems swing doesn't ask for the default exception thread trap so now I have two.
At issue is if we can get a stack trace for the AWT null exception instead of the console message and no trace back. I also turned off all optional LAF's in case it they cause the null exception.
[builder-0.13.b011.jar.zip](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4037167/builder-0.13.b011.jar.zip)
Paul--

**Comment by tddpirate** [comment]
I will try the new jar file in a moment (can you please give the ZIPs some sort of version ID, such as a timestamp?).
Now I have tried something else: I installed `builder-linux-0.13.0.4.tar.gz` and tried to play with the application.
In general, buttons worked. The menu did not work and there were no window decorators (like in the video from an hour ago) so I used again Ctrl-C to kill it.
Buttons and menu items show up only when the mouse goes above them.

Can you test the application in a virtual machine or a Docker container running the xfce4 desktop (preferably under Debian Buster)?

**Comment by tddpirate** [comment]
Hello Paul,
I have reinstalled GUIsliceBuilder 0.13.b011 and copied the jar file from https://github.com/ImpulseAdventure/GUIslice-Builder/issues/57#issuecomment-572256495 (from about half an hour ago).

Apparently the optional LAF's were problematic, because now the application works, is fully responsive, File/Exit works like a charm - and I even was successful in building a trivial project with a button (which changes appearance when touched) and a progress bar (which does nothing at all).

Now I have a version which (partially?) works. and we can further narrow down the bug.

--- Omer Zak

**Comment by Pconti31** [comment]
Great news!  
Ok, tomorrow I'll give it another go and create a version to try and track down the issue further.   
Paul--

**Comment by Pconti31** [comment]
Omar: In thinking about this some more I'm not sure that another version of builder will help much without some more information.   The last version I gave you should allow you to go into
menu items edit->options and change laf one at a time.
The property is called "Theme" and will be set to BusinessBlackSteelSkin.
When you select this field you can scroll up to see standard java laf's.
they are Metal, Nimbus, and CDE/Motif.   The first optional skin after BusinessBlackSteelSkin is
BusinessBlueSteelSkin, select it (we know black doesn't work) and see if we get a crash or NULL exception from the console
Use the terminal console to directly start guislice.sh so we can see console output.
Paul--

**Comment by tddpirate** [comment]
Hello Paul,
As per your request, I experimented with setting Java laf's.
1. Start the application and open Preferences/General dialog. Theme is set to BusinessBlackSteelSkin.
Application works.
2. I change to BusinessBlueSteelSkin.
3. **Application throws `Exception: java.lang.NullPointerException thrown from the UncaughtExceptionHandler in thread "AWT-EventQueue-0"` exceptions. Menu items are not responsive.**
4. **I abort the application using Ctrl-C.**
5. I activate it again. Application is responsive (probably thanks to your patch for turning off optional laf's) and I open Preferences / General. Theme is set to BusinessBlueSteelSkin. Application is responsive.
6. I change the theme to ChallengeDeepSkin.
7. Application behaves as in steps 3,4 above.
8. I activate it again, responsive, theme is set to ChallengerDeepSkin.
9. I change to Metal.
10. Application remains responsive and I close it using File / Exit.
11. I activate it again and change theme to BusinessBlackSteelSkin.
12. Application behaves as in steps 3,4 above.

Attached please find crash log file from step 12 above.
[20200109_172730.txt](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4041105/20200109_172730.txt)

**Comment by tddpirate** [comment]
In one of the experiments, apparently I got a bit different type of crash, without noticing it or noting what I did at the time. The crash log for it is different and I am attaching it just in case it can help you.
[20200109_171642.txt](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4041126/20200109_171642.txt)

**Comment by Pconti31** [comment]
Omar:
Thanks for the crash files.   
Paul--

**Comment by Pconti31** [comment]
Omar:
I made a patch to substance-7.3.1-SNAPSHOT.jar to avoid the null exception,
however this might just cause the same behavior as before.
Keep your #57 jar to go back to in case this fails.   The code in question wasn't written by me and is no longer supported so I might be forced into simply checking for failure and degrading to one of the Java LAF's if my patch fails.

[builder-0.13.b011.jar.zip](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4042299/builder-0.13.b011.jar.zip)
Paul--

**Comment by tddpirate** [comment]
Hello Paul,
The new patch did not work.
Same behavior as in **steps 3,4** above.
--- Omer Zak

**Comment by Pconti31** [comment]
Omar:
First thanks for all the effort you put into this.
This new version will simply attempt to detect failure loading the LAF and if so
degrade to system LAF.   If that fails I may have to put a command line
 switch in to turn off the optional LAFs for people that can't run them.

[builder-0.13.b011.jar-trap.zip](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4042631/builder-0.13.b011.jar-trap.zip)

**Comment by tddpirate** [comment]
The jar from `builder-0.13.b011.jar-trap.zip` (3260724 bytes long, md5sum=`cdfaba85b0cbd3b9f1240c3edbbf347dcdfaba85b0cbd3b9f1240c3edbbf347d`) did not work. Same **steps 3,4** behavior was seen.

**Comment by Pconti31** [comment]
Hello Omar:
Since I can't seem to fix or trap the problem I've added a command line argument that turns off optional LAFs.   Edit your GUIslice.sh script line 75 and change from
`APP_ARGS=$(save "$@")`
to
`APP_ARGS=NO_OPTIONAL_LAFS`
and you should run fine.
This will be added to 0.13.b012.  Your help->about box will id the builder as 0.13.b012-SNAPSHOT-3
and you use that version number for any new bug reports until you can upgrade to 0.13.b012 release.
Thanks again for all your effort!

[builder-0.13.b011.jar-b12-snapshot-3.zip](https://github.com/ImpulseAdventure/GUIslice-Builder/files/4046121/builder-0.13.b011.jar-b12-snapshot-3.zip)

Paul--

**Comment by tddpirate** [comment]
Hello Paul,
I confirm that GUIsliceBuilder with the jar file from `builder-0.13.b011.jar-b12-snapshot-3.zip` works when I invoke `./GUIslice.sh` with the `NO_OPTIONAL_LAFS` argument.
--- Omer Zak

**Comment by Pconti31** [comment]
Thanks again!
I'll be building release 0.13.b012 next monday.  I've decided the easy solution for everyone
is to simply turn off the optional lafs for linux.  So the command line argument won't be needed 
in the future.

**Comment by Pconti31** [comment]
Fixed in Release 0.13.b012
